### PR TITLE
[Fix] Max seen dd

### PR DESCRIPTION
**Pull request by noudcorten**
# Results of merging this

Fixes #

# What has been changed?

* Max seen drawdown calculation based on difference from candle 'open' to candle 'low'
* Changed handling of ROI and stoploss occuring in same candle
* Added more max seen dd tests

# Examples

# Actions to be performed before this can be merged

- [ ] PR depends on #
- [ ] Script to be run

# Security Measures

# Potential Issues / Future considerations

# Assumptions made / Things to add to the wiki
 

# Packages added

# Additional Info

### CODE REVIEW

File: modules/stats/stats.py [hunk 1]
```diff
@@ -252,14 +252,24 @@ def calculate_max_seen_drawdown(self) -> dict:
...
+        open_trade_tick_list = list(open_trade_ticks) if open_trade_ticks is not None else []
+        budget_ticks = timestamp_budget.keys()
+        budget_tick_list = list(budget_ticks) if budget_ticks is not None else []
+        ticks = open_trade_tick_list + budget_tick_list
+        ticks.sort()
+
+        for tick in ticks:
             # Find total value at tick time
-            total_value = get_total_value_at_tick(timestamp_value, timestamp_budget, tick)
+            total_value_low = get_total_value_at_tick(timestamp_value_low, timestamp_budget, tick)
```
Comment (DemianVoorhagen6) [inline modules/stats/stats.py:268]: The function 'get_total_value_at_tick' seems overly complicated, maybe we can refactor it? It takes 3 inputs, returns 3 outputs right away, direct use might even be easier(?)

File: modules/stats/stats.py [hunk 2]
```diff
@@ -252,14 +252,24 @@ def calculate_max_seen_drawdown(self) -> dict:
...
+            total_value_open = get_total_value_at_tick(timestamp_value_open, timestamp_budget, tick)
 
             # Check for new drawdown period
-            if total_value > temp_seen_drawdown['peak']:
+            if total_value_open > temp_seen_drawdown['peak']:
                 # If last drawdown was larger than max drawdown, update max drawdown
                 if temp_seen_drawdown['drawdown'] < max_seen_drawdown['drawdown']:
                     max_seen_drawdown['drawdown'] = temp_seen_drawdown['drawdown']
```
Comment (DemianVoorhagen6) [inline modules/stats/stats.py:275]:
```suggestion
                    max_seen_drawdown = temp_seen_drawdown
```
Not 100% sure this works but believe it does (and all lines beneath can be removed)

File: modules/stats/stats.py [hunk 3]
```diff
@@ -268,17 +278,20 @@ def calculate_max_seen_drawdown(self) -> dict:
...
-                temp_seen_drawdown['drawdown'] = 1.0  # ratio /w respect to peak
+                temp_seen_drawdown['peak'] = total_value_open
+                temp_seen_drawdown['bottom'] = total_value_open
+                temp_seen_drawdown['drawdown'] = 1.0  # ratio w/ respect to peak
                 temp_seen_drawdown['from'] = tick
+
             # Check if drawdown reached new bottom
-            elif total_value < temp_seen_drawdown['bottom']:
-                temp_seen_drawdown['bottom'] = total_value
+            if total_value_low < temp_seen_drawdown['bottom']:
```
Comment (DemianVoorhagen6) [inline modules/stats/stats.py:287]: Shouldn't the new bottom be lower? So > becomes <?

File: modules/stats/trade.py [hunk 4]
```diff
@@ -31,23 +31,29 @@ class Trade:
     sell_reason: SellReason
 
     def __init__(self, ohlcv: dict, spend_amount: float, fee: float, date: datetime, sl_type: str, sl_perc: float):
+        # Basic trade data
         self.status = 'open'
         self.pair = ohlcv['pair']
         self.open = ohlcv['close']
+        self.current = ohlcv['close']
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:38]: Why are .open and .current both referring to close?

File: modules/stats/trade.py [hunk 5]
```diff
@@ -66,13 +72,35 @@ def close_trade(self, reason: SellReason, date: datetime) -> None:
         self.close = self.current
         self.closed_at = date
         self.close_fee_amount = self.capital * self.fee   # final issued fee
+
         self.capital -= self.close_fee_amount
         self.update_profits(update_capital=False)
 
-    def update_stats(self, ohlcv: dict) -> None:
+        # Check if last capital exceeds lowest seen capital (because of issued fee)
+        if self.capital < self.temp_lowest_seen_capital:
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:80]: Why does the variable contain temp? Shouldn't it be current?

File: modules/stats/trade.py [hunk 6]
```diff
@@ -66,13 +72,35 @@ def close_trade(self, reason: SellReason, date: datetime) -> None:
...
+            seen_drawdown = self.capital / self.temp_highest_seen_capital
+            # Check if new seen drawdown exceeds max seen drawdown
+            if seen_drawdown < self.max_seen_drawdown:
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:83]: Shouldn't this be the other way around too? Not sure, fact checking

File: modules/stats/trade.py [hunk 7]
```diff
@@ -66,13 +72,35 @@ def close_trade(self, reason: SellReason, date: datetime) -> None:
...
+        the first time.
+        :type first: bool
+        :return: None
+        :rtype: None
+        """
         self.current = ohlcv['close']
         self.update_profits()
-        self.update_max_drawdown()
+        if not first:
+            self.low_value = ohlcv['low']
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:101]: low_value is ambiguous imo, candle_low would be more suitable if you want to store the candle low. Value is too broad

File: modules/stats/trade.py [hunk 8]
```diff
@@ -66,13 +72,35 @@ def close_trade(self, reason: SellReason, date: datetime) -> None:
...
+            self.open_value = ohlcv['open']
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:102]: Same as above

File: modules/stats/trade.py [hunk 9]
```diff
@@ -92,22 +120,30 @@ def configure_stoploss(self, ohlcv: dict, data_dict: dict) -> None:
...
+        Updates max seen drawdown.
+
+        :return: None
+        :rtype: None
+        """
+        curr_open_capital = self.open_value * self.currency_amount
+        curr_lowest_capital = self.low_value * self.currency_amount
+        
+        # Check for new drawdown period
+        if curr_open_capital > self.temp_highest_seen_capital:
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:133]:
```suggestion
        if curr_open_capital > self.temp_highest_seen_capital:
```
        if curr_total_capital > self.temp_highest_seen_capital:
Open suggests only open positions, total capital/balance/total would be more appropriate

File: modules/stats/trade.py [hunk 10]
```diff
@@ -92,22 +120,30 @@ def configure_stoploss(self, ohlcv: dict, data_dict: dict) -> None:
...
+            # Reset temp seen drawdown stats
+            self.temp_highest_seen_capital = curr_open_capital
+            self.temp_lowest_seen_capital = curr_open_capital
+            self.temp_seen_drawdown = 1.0   # ratio w/ respect to peak
+
```
Comment (DemianVoorhagen6) [inline modules/stats/trade.py:138]: I've seen this kind of implementation a lot in this PR, maybe make a function/one-liner to do this. Would save tens of lines in these files already.

File: modules/stats/tradingmodule.py [hunk 11]
```diff
@@ -28,48 +28,74 @@ def __init__(self, config: TradingModuleConfig):
...
     def close_trade(self, trade: Trade, reason: SellReason, ohlcv: dict) -> None:
         date = datetime.fromtimestamp(ohlcv['time'] / 1000)
         trade.close_trade(reason, date)
 
-        self.total_fee_amount += trade.close_fee_amount
+        if trade.sell_reason == SellReason.STOPLOSS_AND_ROI:
+            # Because trade had no impact on results, remove first issued fee from
+            # total amount of fee and reset trade stats.
+            first_fee = trade.starting_amount * trade.fee
+            self.total_fee_amount -= first_fee
```
Comment (DemianVoorhagen6) [inline modules/stats/tradingmodule.py:91]:
```suggestion
            self.total_fee_amount -= first_fee
```
            self.total_fee_paid -= first_fee

File: modules/stats/tradingmodule.py [hunk 12]
```diff
@@ -28,48 +28,74 @@ def __init__(self, config: TradingModuleConfig):
...
+
+            # Reset trade stats
```
Comment (DemianVoorhagen6) [inline modules/stats/tradingmodule.py:93]: Does this cover the un-do of the trade on all levels? (Portfolio, coins, trade level)
Does .remove(trade) also reduce the amount of trades shown at the end?

File: modules/stats/tradingmodule.py [hunk 13]
```diff
@@ -97,17 +123,19 @@ def open_trade(self, ohlcv: dict, data_dict: dict) -> None:
...
         # Update total budget with configured spend amount and fee
         self.total_fee_amount += spend_amount * self.fee
         self.budget -= spend_amount
         self.open_trades.append(new_trade)
-        self.update_value_per_timestamp_tracking(new_trade, ohlcv)
+        self.update_open_trades_value_per_timestamp(new_trade, ohlcv)
 
     def check_roi_open_trade(self, trade: Trade, ohlcv: dict) -> bool:
         time_passed = datetime.fromtimestamp(ohlcv['time'] / 1000) - trade.opened_at
         profit_percentage = ((ohlcv['high'] / trade.open) - 1.) * 100
```
Comment (DemianVoorhagen6) [inline modules/stats/tradingmodule.py:136]: The roi isn't necessarily 'high'?

File: modules/stats/tradingmodule.py [hunk 14]
```diff
@@ -135,20 +163,40 @@ def find_open_trade(self, pair: str) -> Optional[Trade]:
...
+        """
+        Method is used to be able to track the open trades value per timestamp
+        """
         trade_opened_at = datetime.timestamp(trade.opened_at) * 1000
         if trade_opened_at == ohlcv['time']:
-            try:
-                self.open_order_value_per_timestamp[ohlcv['time']] += trade.starting_amount
-            except KeyError:
-                self.open_order_value_per_timestamp[ohlcv['time']] = trade.starting_amount
+            self.open_trades_open_per_timestamp[ohlcv['time']] = \
```
Comment (DemianVoorhagen6) [inline modules/stats/tradingmodule.py:172]: Feels a bit though to read, no concrete solution comes to mind though

File: modules/stats/tradingmodule.py [hunk 15]
```diff
@@ -135,20 +163,40 @@ def find_open_trade(self, pair: str) -> Optional[Trade]:
...
         else:
-            try:
-                self.open_order_value_per_timestamp[ohlcv['time']] += trade.capital
-            except KeyError:
-                self.open_order_value_per_timestamp[ohlcv['time']] = trade.capital
-
-    def update_budget_per_timestamp_tracking(self, ohlcv: dict) -> None:
+            # When trade.low_value is equal to trade.current in final candle, trade.capital could 
+            # be lower than curr_low_capital due to fee.
+            curr_low_capital = trade.low_value * trade.currency_amount
```
Comment (DemianVoorhagen6) [inline modules/stats/tradingmodule.py:179]:
```suggestion
            curr_low_capital = trade.low_value * trade.currency_amount
```
Is low value for the lowest value? If so trough would be more meaningfull, e.g.:
curr_low_capital = trade.trough_value * trade.currency_amount
If this also applies for trough capital:
curr_trough_capital = trade.trough_value * trade.currency_amount

File: modules/stats/tradingmodule.py [hunk 16]
```diff
@@ -135,20 +163,40 @@ def find_open_trade(self, pair: str) -> Optional[Trade]:
...
+            low_capital = curr_low_capital if curr_low_capital < trade.capital \
+                else trade.capital
+            self.open_trades_low_per_timestamp[ohlcv['time']] = \
```
Comment (DemianVoorhagen6) [inline modules/stats/tradingmodule.py:182]: open_trades_low_per_timestamp is multi interpretable

File: test/stats/test_coin_insights.py [hunk 17]
```diff
@@ -166,3 +166,107 @@ def test_profit():
...
+    stats = fixture.create().analyze()
+
+    # Assert
+    assert math.isclose(stats.coin_res[0].max_seen_drawdown, -75.25)
+    assert math.isclose(stats.coin_res[0].max_realised_drawdown, -50.995)
+
+
+def test_drawdown_multiple_trades():
+    """Given multiple trades, 'max seen drawdown' should be the lowest seen drawdown
+    of the combined trades"""
```
Comment (DemianVoorhagen6) [inline test/stats/test_coin_insights.py:229]: Can you brief me on this one?

File: test/stats/test_roi_stoploss.py [hunk 18]
```diff
@@ -22,8 +22,89 @@ def test_roi():
     assert stats.main_results.end_capital == 245.025
 
 
+def test_stoploss():
+    """Given 'value of coin falls below stoploss', 'profit' should 'lower same minus fee'"""
```
Comment (DemianVoorhagen6) [inline test/stats/test_roi_stoploss.py:26]: should be

**Review by DemianVoorhagen6 — CHANGES_REQUESTED** [review]
Lets go over the comments f2f after you read everything, I think we could improve some more.

File: modules/stats/tradingmodule.py [hunk 19]
```diff
@@ -28,48 +28,74 @@ def __init__(self, config: TradingModuleConfig):
...
+        Method is called when specified pair has open trades.
+        checks for ROI
+        checks for stoploss
+        :param ohlcv: dictionary with OHLCV data for current tick
+        :type ohlcv: dict
+        :param trade: Trade corresponding to tick pair
+        :type trade: Trade
+        :return: None
+        :rtype: None
+        """
```
Comment (sheridavandenbent) [inline modules/stats/tradingmodule.py:65]:
```suggestion
```

Remove all unneccesary comments!

File: modules/stats/trade.py [hunk 20]
```diff
@@ -65,14 +71,36 @@ def close_trade(self, reason: SellReason, date: datetime) -> None:
...
+        Updates states according to latest data.
+
+        :param ohlcv: dictionary with OHLCV data for current tick
+        :type ohlcv: dict
+        :param first: boolean which indicates whether the function is called for
+        the first time.
+        :type first: bool
+        :return: None
+        :rtype: None
+        """
```
Comment (sheridavandenbent) [inline modules/stats/trade.py:97]:
```suggestion
```

Unnecessary comment!

File: modules/stats/trade.py [hunk 21]
```diff
@@ -92,22 +109,31 @@ def configure_stoploss(self, ohlcv: dict, data_dict: dict) -> None:
...
-    def set_drawdown_if_lower(self):
-        if self.temp_seen_drawdown_price < self.max_seen_drawdown_price:
-            self.max_seen_drawdown_price = self.temp_seen_drawdown_price
-            self.max_seen_drawdown = self.temp_seen_drawdown_price / self.seen_peak_capital
+        """
+        Updates max seen drawdown defined as: biggest difference between
+        highest peak of candle 'open' and lowest bottom of candle 'low'.
+
+        :return: None
+        :rtype: None
```
Comment (sheridavandenbent) [inline modules/stats/trade.py:117]:
```suggestion
```

Unnecessary comments

**Review by sheridavandenbent — APPROVED** [review]

**Review by DemianVoorhagen6 — DISMISSED** [review]
Better :) I did notice new/other variable ambiguities, so let's be extra watchful on those to make sure we maintain high code quality and keep the project scalable.

**Review by sheridavandenbent — APPROVED** [review]

**Pull request merged**